### PR TITLE
HDDS-10889. Remove certificate revocation related code.

### PR DESCRIPTION
**Pull request by fapifta**
## What changes were proposed in this pull request?
In this PR, I propose to remove and stop maintaining Certificate Revocation related code in its current form.

The main reason behind this move is that it is not finished, it does not work currently as it misses some things still, and it is unreachable for any regular user.

The secondary reason to remove this code from our codebase is that, for CRL distribution there are simple mechanisms in the SSL layer that we can use. There is the CRL distribution endpoint that can be encoded to the certificates, and CRL checks can be pushed down to the SSL layer for communication.

Moreover, the current code still use these certs to verify the signature of Delegation Tokens but that is something we are planning to change to use symmetric encryption similarly to block and container tokens, as of now the JCE code does not check the CRL Distribution endpoint when it verifies a signature, but that is exactly the same case as we have today, so functionally the system does not degrade by this move.
There is one more reason though, and that is the burden on maintaining this code part in the compliance mode related development, as it has some hard to replace bouncycastle usages also.

With all that, as we still use protoc2.5 and we don't have a safe option to reserve fields that were once declared and prevent accidental identifier re-use, I kept the relevant declarations in our proto files, with some comments.
Ideally in a protoc3 world we can use the reserved keyword for unused fields and just remove them from messages, or remove (with that empty) all messages that are affected. But as of now we are compiling quite a few things with protoc 2.5, because if we do use a different gRPC base Java runtime, as it would not fly with a protoc3 compiled version, as the RPC libraries in the Hadoop2 filesystem lib are not able to use the protoc3 compiled stuff without the proto 3 runtime, which we can not really enforce to be loeaded by hadoop2 properly.
After we migrate to proto3 we can do the proper removal of the proto fields, so this patch does not want to change proto definitions just mark them.

The resulting code, if it meets with an old client that tries to send extra field data that is not processed anywhere, it is free to do it, as the protoc parsing will either read those fileds  and our code won't use it (with proto 2.5.), or it will read it as an unknown field (with the reserved tags provided by prtoc 3 and suggested by the comments.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-10889

## How was this patch tested?

No new logic added, and the removed code is dead code, current tests should cover.

### CODE REVIEW

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMCertStore.java [hunk 1]
```diff
@@ -129,98 +108,15 @@ public void storeValidScmCertificate(BigInteger serialID,
   public void checkValidCertID(BigInteger serialID) throws IOException {
     lock.lock();
     try {
-      if ((getCertificateByID(serialID, VALID_CERTS) != null) ||
-          (getCertificateByID(serialID, CertType.REVOKED_CERTS) != null)) {
+      if ((getCertificateByID(serialID) != null) ||
```
Comment (Galsza) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMCertStore.java:111]: Minor: we are repeating the same check twice here

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMSecurityProtocolServer.java [hunk 2]
```diff
@@ -439,20 +424,18 @@ public String getCACertificate() throws IOException {
...
    */
   @Override
   public List<String> listCertificate(NodeType role,
-      long startSerialId, int count, boolean isRevoked) throws IOException {
+      long startSerialId, int count) throws IOException {
     List<X509Certificate> certificates =
-        scmCertificateServer.listCertificate(role, startSerialId, count,
-            isRevoked);
+        scmCertificateServer.listCertificate(role, startSerialId, count
+        );
```
Comment (Galsza) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMSecurityProtocolServer.java:438]: unnecessary new line here

**Review by Galsza — APPROVED** [review]
Thank you @fapifta for your changes, I've left two minor comments.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMSecurityProtocolServer.java (hunk 2 above)
Comment (fapifta) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMSecurityProtocolServer.java]: you are right, thank you for spotting this one...

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMCertStore.java (hunk 1 above)
Comment (fapifta) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMCertStore.java]: you are right, thank you for spotting this one...

**Comment by fapifta** [comment]
Thank you for the review @Galsza if you spot any other issue, please feel free to note it, I have addressed the two comments.

**Comment by kerneltime** [comment]
FYI: This feature is tracked in https://issues.apache.org/jira/browse/HDDS-7333
LGTM.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/MiniOzoneHAClusterImpl.java [hunk 3]
```diff
@@ -661,7 +661,6 @@ private void initSCMHAConfig() {
         conf.set(scmHttpAddrKey, localhostWithFreePort());
         conf.set(scmHttpsAddrKey, localhostWithFreePort());
         conf.set(scmSecurityAddrKey, localhostWithFreePort());
-        conf.set("ozone.scm.update.service.port", "0");
```
Comment (kerneltime) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/MiniOzoneHAClusterImpl.java:664]: Is this related to this PR?

**Comment by Galsza** [comment]
LGTM +1

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/authority/CertificateStore.java [hunk 4]
```diff
@@ -74,40 +66,8 @@ void storeValidScmCertificate(BigInteger serialID,
...
-                                    CRLApprover approver)
-      throws IOException;
-
-  /**
-   * Deletes an expired certificate from the store. Please note: We don't
-   * remove revoked certificates, we need that information to generate the
-   * CRLs.
-   * @param serialID - Certificate ID.
-   */
-  void removeExpiredCertificate(BigInteger serialID) throws IOException;
```
Comment (dombizita) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/authority/CertificateStore.java:107]: Why did you remove this too? Also you didn't remove `removeAllExpiredCertificates` below.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMSecurityProtocolServer.java [hunk 5]
```diff
@@ -439,28 +424,23 @@ public String getCACertificate() throws IOException {
...
+    List<X509Certificate> certificates = scmCertificateServer.listCertificate(role, startSerialId, count);
     List<String> results = new ArrayList<>(certificates.size());
     for (X509Certificate cert : certificates) {
       try {
         String certStr = getPEMEncodedString(cert);
         results.add(certStr);
       } catch (SCMSecurityException e) {
-        throw new SCMSecurityException("listCertificate operation failed.",
-            e, e.getErrorCode());
+        throw new SCMSecurityException("listCertificate operation failed.", e, e.getErrorCode());
```
Comment (dombizita) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMSecurityProtocolServer.java:443]: NIT: unnecessary change

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/MiniOzoneHAClusterImpl.java (hunk 3 above)
Comment (fapifta) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/MiniOzoneHAClusterImpl.java:664]: Hi @kerneltime,

yes it is related, the CRL distribution mechanism uses the SCMUpdateService (actually the CRL distribution is the only thing uses it), and it has a config object UpdateServiceConfig which defines this config.
As in this PR I am removing the update service as it is not used by anything else, it is reasonable to remove the config option and its usages also.

File: hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/TestHddsDatanodeService.java [hunk 6]
```diff
@@ -106,22 +103,6 @@ public void setUp() throws IOException {
...
-    service.start(conf);
-
-    assertNotNull(service.getDatanodeDetails());
-    assertNotNull(service.getDatanodeDetails().getHostName());
-    assertFalse(service.getDatanodeStateMachine().isDaemonStopped());
-    assertNotNull(service.getCRLStore());
-
-    service.stop();
-    // CRL store must be stopped when the service stops
-    assertNull(service.getCRLStore().getStore());
```
[the remark this answers is not in the document]
Comment (fapifta) [reply hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/TestHddsDatanodeService.java:120]: Hi @dombizita,

first I just removed the CRL related part, then I took a look at the testcase, and found that it calls only the start, and stop method of the HDDSDataNodeService class, and in between it does 3 assertions:
1. the DatanodeDetails is not null
2. the DatanodeDetails contains a hostname that is not null
3. the Datanode's statemachine is not stopped.

From these three assertions, none seems to be useful as if these assertions are not correct, then something should fail separately as these as far as I can tell fundamental things that are needed for normal operations also, so as a last resort our docker based tests should fail on read and write operations if these assertions are not true, hence I propose to remove the remaining part of the test also.

File: hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/common/report/TestReportPublisher.java [hunk 7]
```diff
@@ -182,66 +170,4 @@ public void testCommandStatusPublisher() throws InterruptedException {
...
-    }
-    executorService.shutdown();
-  }
-
-  /**
-   * Get a datanode details.
-   *
-   * @return DatanodeDetails
-   */
-  private static DatanodeDetails getDatanodeDetails() {
```
[the remark this answers is not in the document]
Comment (fapifta) [reply hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/common/report/TestReportPublisher.java:225]: yes, however I think it does worth to remove it right now, and do not separate it to an other PR if I already touch this class.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/protocol/SCMSecurityProtocol.java [hunk 8]
```diff
@@ -110,14 +109,13 @@ String getSCMCertificate(ScmNodeDetailsProto scmNodeDetails,
   /**
    * Get list of certificates meet the query criteria.
    *
-   * @param type            - node type: OM/SCM/DN.
-   * @param startSerialId   - start certificate serial id.
-   * @param count           - max number of certificates returned in a batch.
-   * @param isRevoked       - whether list for revoked certs only.
+   * @param type          - node type: OM/SCM/DN.
+   * @param startSerialId - start certificate serial id.
+   * @param count         - max number of certificates returned in a batch.
```
[the remark this answers is not in the document]
Comment (fapifta) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/protocol/SCMSecurityProtocol.java:114]: Yes, you are right, we should leave these whitespaces as it is, however I have pulled the method definition to be in one line as with new line length limits it fits.

File: hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/TestHddsDatanodeService.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (fapifta) [reply hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/TestHddsDatanodeService.java:120]: Though, looking at it once more, I think it is best to move these three assertions to the test that remains in this class, just to be sure, I hope you are fine with it.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/authority/CertificateStore.java (hunk 4 above)
Comment (fapifta) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/authority/CertificateStore.java:107]: This method is defined, and it has two implementation, one in MockCAStore, and SCMCertStore, neither of them did anything meaningful, and actually removing an expired certificate by ID does not seem to be useful even in the future. It is not closely related to CRL, but I ran into this while I was wondering around the code looking at what to remove for CRL.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMSecurityProtocolServer.java (hunk 5 above)
Comment (fapifta) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMSecurityProtocolServer.java:443]: I have changed this line, as there was already a change in this method, so it felt natural to reformat the method body, as this can fit into the new line length limit, and with that there won't be wrapping anymore in the method. I can change this back if you want though.

**Review by kerneltime — APPROVED** [review]

**Review by adoroszlai — APPROVED** [review]

**Review by dombizita — APPROVED** [review]

**Comment by dombizita** [comment]
Thanks for working on this @fapifta! Thanks for the reviews @adoroszlai, @kerneltime and @Galsza!
